src/pytorch_lightning/core/module.py

```diff
@@ -1950,6 +1951,70 @@ def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
                 weakref.proxy(self), pre_load_state_dict_hook, True  # type: ignore[arg-type]
             )
 
+    @classmethod
+    def from_compiled(cls, model: Union["pl.LightningModule", "torch._dynamo.OptimizedModule"]) -> "pl.LightningModule":
+        """Returns an instance LightningModule from the output of ``torch.compile``.
+
+        The ``torch.compile`` function returns a ``torch._dynamo.OptimizedModule``, which wraps the LightningModule
+        passed in as an argument, but doesn't inherit from it. This means that the output of ``torch.compile`` behaves
+        like a LightningModule but it doesn't inherit from it (i.e. `isinstance` will fail).
+
+        Use this method to obtain a LightningModule that still runs with all the optimizations from ``torch.compile``.
+        """
+
+        from torch._dynamo import OptimizedModule
+
+        if not isinstance(model, OptimizedModule):
+            raise ValueError(
+                "`model` is required to be a `torch._dynamo.OptimizedModule`. " f"Found a `{type(model)}` instead."
+            )
+
+        orig_module = model._orig_mod
+
+        if not isinstance(orig_module, cls):
+            raise ValueError(
+                "`model` is expected to be a compiled LightingModule. " f"Found a compiled {type(orig_module)} instead"
+            )
+
+        orig_module._compiler_ctx = {
+            "compiler": "dynamo",
+            "dynamo_ctx": model.dynamo_ctx,
+            "original_forward": orig_module.forward,
+        }
+
+        orig_module.forward = model.dynamo_ctx(orig_module.forward)  # type: ignore[assignment]
```

Comment:
Let's wrap `training_step`, some users might not use forward in their model.

Reply:
of course, that makes sense

Reply:
I'm kind of hesitant because training_step contains things that dynamo doesn't seem to be tested on right now. I propose we merge this now and work on wrapping training_step later as the technology matures